jep/0001/README.adoc

```diff
@@ -0,0 +1,306 @@
+eventual consistency management.
+This layer will be determined during reference implementation polishing.
+
+=== Pipeline Bridge extensions
+
+The plugin should also implement Pipeline `LogStorage` and `LogStorageFactory` extension points
+so that it transparently supports Pipeline with existing API.
+
+Pipeline Storage JEP is documented in JEP-TODO.
+
+=== JSON-based external-logging layer
+
+The API Plugin should offer a convenience layer in order
+to support a number of most common logging providers like
+Logstash, Fluentd, Elasticsearch, AWS CloudWatch, etc.
+
+This layer should provide the following features:
+
+* Utility classes for reading and writing JSON Events,
+including parsing/writing `ConsoleNote` objects
+* Base classes for constructing JSON queries for fetching data
+
+Other convenience layers will be defined during prototyping.
+
+=== Configuration
+
+External Logging API plugin should be fully configurable via
+WebUI and _Configuration-as-Code Plugin_.
+It includes:
+
+* Selection of `LoggingMethod` and `LogStorage` factories
+* Configuration of built-in External Logging and Log Browser factories
+* Any other configuration options
+
+=== Reference implementation
+
+As a reference implementation of the _External Logging API_ plugin,
+a new _External Logging for Elasticsearch Plugin_ will be implemented.
+Other implementations may be also created.
+
+== Reasoning
+
+=== Why separate plugin?
+
+One may say that all bits in the current design could be implemented
+as a part of the Jenkins core.
+It is true, but detaching of the plugin is intentional:
+
+* The plugin will have a separate release cycle so that changes in it
+  can be delivered and backported independently from the core's release cycle
+* The approach allows keeping the patches on the core's side minimal
+
+All External Log Storage implementations are expected to extend this plugin
+instead of just using API provided by the Core.
+Core APIs may be still used to define custom `LoggingMethodLocator` impelemtations,
+e.g. to define a custom logger allocation logic.
+
+=== Why do we need JSON layer?
+
+Many popular log storage engines store events in a JSON format:
+Fluentd, Logstash, Elasticsearch, AWS CloudWatch, etc.
+Offering a JSON layer as a part of the API plugin could greatly
+simplify such implementations.
+
+=== Eventually-consistent storages
+
+Some target storages are eventually consistent.
+One cannot just write the data to remote storage and then reliably read it.
+It is critical for log browsers:
+
+* When a run finishes, querying data does not guarantee we get all the data
+* `Loggable#isCompleted()` call is not enough,
+some entries may be missing for "completed" entries
+
+Such issue explains why we need a special watch layer to determine whether logs are actually completed.
+It explains why we need a special logic/queries to determine whether the log is actually complete.
+
+Full logic for the layer should be determined during reference implementation.
```

Comment:
BTW this is solved for the Pipeline implementation.

Reply:
For sure.